src/crush/CrushWrapper.cc

```diff
@@ -1035,6 +1035,15 @@ int CrushWrapper::detach_bucket(CephContext *cct, int item)
   return bucket_weight;
 }
 
+bool CrushWrapper::check_bucket_relation(crush_bucket *f, crush_bucket *tar){
```

Comment:
instead of going with the recursive depth-first search, could just check if the bucket `a` has an ancestor of `b`, like:
```c
bool is_parent_of(int child, int p) {
  int parent = 0;
  while (!get_immediate_parent_id(child, &parent)) {
    if (parent == p) {
      return true;
    }
    child = parent;
  }
  return false;
}
```